src/main/java/org/elasticsearch/gateway/GatewayAllocator.java

```diff
@@ -142,28 +154,70 @@ public boolean allocateUnassigned(RoutingAllocation allocation) {
                     continue;
                 }
                 if (version != -1) {
+                    // We found one that can be allocated!
```

Comment:
the logic here becomes complicated, I think we can simplify it. In `buildShardStates`, we can remove the special handling of fulling the system into version 0 for `recvoerOnAnyNode`, and have a simple if check on `recoverOnEveryNode`, if it is, just capture the version, and highest, and if not, do the regular logic?